src/systems/collision.rs

```diff
@@ -0,0 +1,180 @@
+                }
+            }
+        }
+    }
+}
+
+pub struct CollisionHandlerForObstacles;
+
+impl<'s> System<'s> for CollisionHandlerForObstacles {
+    type SystemData = (
+        ReadExpect<'s, Collisions>,
+        WriteStorage<'s, Transform>
+        );
+
+    fn run(&mut self, (collisions, mut transforms): Self::SystemData) {
+        for collision in &collisions.0 {
+            if let Some(transform) = transforms.get_mut(collision.collider)
+            {
+                transform.prepend_translation_x(-collision.vector.x);
+                transform.prepend_translation_y(-collision.vector.y);
+            }
+        }
+
+    }
+}
+
+pub struct ProjectileCollisionSystem;
+
+impl<'s> System<'s> for ProjectileCollisionSystem {
+    type SystemData = (
+        WriteStorage<'s, Transform>,
+        ReadStorage<'s, Projectile>,
+        ReadStorage<'s, BoundingCircle>,
+        Entities<'s>,
+        WriteExpect<'s, ProjectileCollisions>
+        );
+    fn run(&mut self, (transforms, projectiles, bounding_circles, entities, mut collision_resource): Self::SystemData) {
+        collision_resource.0.clear();
+        for projectile in (&transforms, &projectiles, &entities).join()
+        {
+            for object in (&transforms, &bounding_circles, &entities).join()
+            {
+                // unlikely
+                if projectile.2 == object.2
+                {
+                    continue;
+                }
+
+                if let Some(collision) = check_projectile_collision(
+                    projectile.0,
+                    Collider{transform: object.0, bound: object.1})
+                {
+                    collision_resource.0.push(ProjectileCollision{
+                        projectile: projectile.2,
+                        target: object.2,
+                        vector: collision});
+                }
+            }
+        }
+    }
+}
+
+pub struct ProjectileCollisionHandler;
+
+impl<'s> System<'s> for ProjectileCollisionHandler {
+    type SystemData = (
+        Entities<'s>,
+        ReadExpect<'s, ProjectileCollisions>
+        );
+
+    // Here Projectile components are not explicitly filtered. ProjectCollisionSystem is expected
+    // to put proper entities in `collision.projectile`
+    fn run(&mut self, (entities, collisions): Self::SystemData) {
+        for collision in &collisions.0 {
+            entities.delete(collision.projectile).expect("JAJAJa");
+        }
+
+    }
+}
+
+struct Collider<'a>
+{
+    transform: &'a Transform,
+    bound: &'a BoundingCircle,
+}
+
+fn calculate_disposition(a: &Transform, b: &Transform) -> Vector2<f32>
+{
+    Vector2::new(
+        b.translation().x - a.translation().x,
+        b.translation().y - a.translation().y
+        )
+}
+
+fn calculate_length(vec: &Vector2<f32>) -> f32
+{
+    let mut distance = (vec.x.powf(2.0) + vec.y.powf(2.0)).sqrt();
+    if distance == 0.0
```

Comment:
Exact matching against float might give funky results. Not sure about the required behavior of this function, but consider the following:
```
distance.norm().max(0.0001)
```

Reply:
The tests written in the meantime showed that this implementation is still not sufficient. (The disposition vector cannot be normailzed if its about to be zero), so I had to handle the "close to zero" case specially. I have deliberately chosen a direction to the collision vector, and the treshold (0.0001) is still subject to debate.